File: client-apps/cli/internal/cli/apply/synthesize.go
// Package apply provides SDK synthesis execution for the stigmer apply command.
//
// The synthesizer executes user SDK programs (Go, Python, Node) and collects
// the generated resource manifests for deployment to the backend.
//
// SDK Execution Model:
// Users write declarative code using the Stigmer SDK that registers resources
// (agents, skills, workflows). The CLI executes the program with STIGMER_OUT_DIR
// set, and the SDK writes individual .pb files for each resource.
//
// Supported Runtimes:
//   - Go: `go run <entry_point>`
//   - Python: `python <entry_point>`
//   - Node: `npx ts-node <entry_point>` for .ts, `node <entry_point>` for .js
package apply

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/pkg/errors"
	"github.com/stigmer/stigmer/client-apps/cli/internal/cli/synthesis"
)

// SynthesizeOptions configures the SDK synthesis execution.
type SynthesizeOptions struct {
	// ProjectDir is the directory containing the SDK code and stigmer.yaml.
	ProjectDir string

	// Runtime specifies which language runtime to use (go, python, node).
	Runtime Runtime

	// EntryPoint is the SDK entry point file (e.g., main.go, main.py, index.ts).
	EntryPoint string

	// Quiet suppresses SDK stdout output during synthesis.
	Quiet bool
}

// SynthesizeResult contains the outcome of SDK synthesis.
type SynthesizeResult struct {
	// OutputDir is the directory containing synthesized .pb files.
	OutputDir string

	// Result contains parsed resource manifests from synthesis output.
	Result *synthesis.Result

	// Stdout contains SDK program stdout (useful for debugging).
	Stdout string
}

// Synthesize executes the SDK entry point and captures synthesized manifests.
//
// The function:
// 1. Validates inputs (project dir exists, entry point exists)
// 2. Creates .stigmer/ output directory
// 3. Runs runtime-specific preparation (e.g., go mod tidy)
// 4. Executes SDK with STIGMER_OUT_DIR environment variable
// 5. Parses synthesized .pb files using synthesis.ReadFromDirectory()
//
// Returns a SynthesizeResult containing parsed manifests, or an error with
// actionable guidance for common issues (missing dependencies, compile errors).
func Synthesize(opts *SynthesizeOptions) (*SynthesizeResult, error) {
	if opts == nil {
		return nil, errors.New("synthesize options cannot be nil")
	}

	// Validate project directory exists
	if opts.ProjectDir == "" {
		return nil, errors.New("project directory is required")
	}

	absProjectDir, err := filepath.Abs(opts.ProjectDir)
	if err != nil {
		return nil, errors.Wrap(err, "failed to resolve project directory")
	}

	if _, err := os.Stat(absProjectDir); os.IsNotExist(err) {
		return nil, errors.Errorf("project directory not found: %s", opts.ProjectDir)
	}

	// Validate entry point exists
	if opts.EntryPoint == "" {
		return nil, errors.New("entry point is required")
	}

	entryPointPath := filepath.Join(absProjectDir, opts.EntryPoint)
	if _, err := os.Stat(entryPointPath); os.IsNotExist(err) {
		return nil, errors.Errorf("entry point not found: %s", opts.EntryPoint)
	}

	// Validate runtime
	if opts.Runtime == "" {
		return nil, errors.New("runtime is required (go, python, or node)")
	}

	// Create output directory
	outputDir := filepath.Join(absProjectDir, ".stigmer")
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return nil, errors.Wrap(err, "failed to create output directory")
	}

	// Run runtime-specific preparation
	if err := prepareRuntime(opts.Runtime, absProjectDir); err != nil {
		return nil, err
	}

	// Build and execute the SDK command
	cmdArgs, err := getRuntimeCommand(opts.Runtime, opts.EntryPoint)
	if err != nil {
		return nil, err
	}

	cmd := exec.Command(cmdArgs[0], cmdArgs[1:]...)
	cmd.Dir = absProjectDir
	cmd.Env = append(os.Environ(), "STIGMER_OUT_DIR="+outputDir)

	// Capture stdout and stderr
	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		errMsg := formatExecutionError(opts.Runtime, stderr.String(), err)
		return nil, errors.New(errMsg)
	}

	// Parse synthesized manifests
	result, err := synthesis.ReadFromDirectory(outputDir)
	if err != nil {
		return nil, errors.Wrap(err, "failed to read synthesis output")
	}

	return &SynthesizeResult{
		OutputDir: outputDir,
		Result:    result,
		Stdout:    stdout.String(),
	}, nil
}

// getRuntimeCommand returns the command arguments for executing the SDK entry point.
func getRuntimeCommand(runtime Runtime, entryPoint string) ([]string, error) {
	switch runtime {
	case RuntimeGo:
		return []string{"go", "run", entryPoint}, nil

	case RuntimePython:
		if _, err := exec.LookPath("python3"); err == nil {
			return []string{"python3", entryPoint}, nil
		}
		return []string{"python", entryPoint}, nil

	case RuntimeNode:
		ext := filepath.Ext(entryPoint)
		if ext == ".ts" || ext == ".tsx" || ext == ".mts" {
			return []string{"npx", "ts-node", entryPoint}, nil
		}
		return []string{"node", entryPoint}, nil

	default:
		return nil, errors.Errorf("unsupported runtime: %v", runtime)
	}
}

// prepareRuntime runs runtime-specific preparation steps before SDK execution.
func prepareRuntime(runtime Runtime, projectDir string) error {
	switch runtime {
	case RuntimeGo:
		return prepareGoRuntime(projectDir)
	case RuntimePython:
		return preparePythonRuntime(projectDir)
	case RuntimeNode:
		return prepareNodeRuntime(projectDir)
	default:
		return nil
	}
}

// prepareGoRuntime ensures Go dependencies are up to date.
func prepareGoRuntime(projectDir string) error {
	// Check if go.mod exists
	goModPath := filepath.Join(projectDir, "go.mod")
	if _, err := os.Stat(goModPath); os.IsNotExist(err) {
		return errors.New("go.mod not found: run 'go mod init' to initialize a Go module")
	}

	// Run go mod tidy to ensure dependencies are current
	cmd := exec.Command("go", "mod", "tidy")
	cmd.Dir = projectDir
	cmd.Env = os.Environ()

	output, err := cmd.CombinedOutput()
	if err != nil {
		return errors.Errorf("failed to update Go dependencies:\n%s\n\nRun 'go mod tidy' manually to resolve", string(output))
	}

	return nil
}

// preparePythonRuntime validates Python environment is ready.
func preparePythonRuntime(projectDir string) error {
	// Check if python/python3 is available
	pythonCmd := "python"
	if _, err := exec.LookPath("python3"); err == nil {
		pythonCmd = "python3"
	} else if _, err := exec.LookPath("python"); err != nil {
		return errors.New("Python not found: install Python 3.x and ensure it's in your PATH")
	}

	// Verify Python version is 3.x
	cmd := exec.Command(pythonCmd, "--version")
	output, err := cmd.Output()
	if err == nil && !strings.Contains(string(output), "Python 3") {
		return errors.New("Python 3.x required: your system has Python 2.x")
	}

	// Check for requirements.txt and suggest install if present
	reqPath := filepath.Join(projectDir, "requirements.txt")
	if _, err := os.Stat(reqPath); err == nil {
		// requirements.txt exists - user should install deps
		// We don't auto-install to avoid modifying user's environment
		// The error will be caught during execution if deps are missing
	}

	return nil
}

// prepareNodeRuntime validates Node.js environment is ready.
func prepareNodeRuntime(projectDir string) error {
	// Check if node is available
	if _, err := exec.LookPath("node"); err != nil {
		return errors.New("Node.js not found: install Node.js and ensure it's in your PATH")
	}

	// Check for package.json
	pkgPath := filepath.Join(projectDir, "package.json")
	if _, err := os.Stat(pkgPath); os.IsNotExist(err) {
		return errors.New("package.json not found: run 'npm init' to initialize a Node.js project")
	}

	// Check if node_modules exists (deps installed)
	modulesPath := filepath.Join(projectDir, "node_modules")
	if _, err := os.Stat(modulesPath); os.IsNotExist(err) {
		return errors.New("node_modules not found: run 'npm install' to install dependencies")
	}

	return nil
}

// formatExecutionError creates a helpful error message based on runtime and error output.
func formatExecutionError(runtime Runtime, stderr string, execErr error) string {
	if len(stderr) > 800 {
		stderr = stderr[:800] + "\n... (truncated)"
	}

	var guidance string
	switch runtime {
	case RuntimeGo:
		guidance = "Check for compile errors above. Run 'go build' to see full error output."
	case RuntimePython:
		guidance = "If you see import errors, run 'pip install -r requirements.txt' in a virtual environment."
	case RuntimeNode:
		guidance = "If you see module errors, run 'npm install' to install dependencies."
	}

	if stderr == "" {
		stderr = execErr.Error()
	}

	return "SDK synthesis failed:\n" + stderr + "\n\n" + guidance
}

File: client-apps/cli-ts/src/resources/apply/synth/synthesize.ts
// SDK synthesis execution.
//
// Runs a project's `entry_point` as a subprocess with `STIGMER_OUT_DIR` (and
// `STIGMER_ORG_ID`) set; the program writes `.pb` files into that dir, which the
// reader then decodes. Uniform subprocess model for every runtime (DD-009 §5):
//   go      → `go run <entry>`
//   python  → `python3 <entry>`
//   node    → `npx tsx <entry>` for TS, `node <entry>` for JS
//
// `tsx` (not Go's `ts-node`) executes TypeScript entries — it is the runner the
// cli-ts ecosystem already standardizes on (DD-009 alternatives). The spawn is
// injected so the orchestration is unit-testable without a real subprocess
// (mirrors `connect/oauth.ts`). Port of `internal/cli/apply/synthesize.go`.

import { spawn } from "node:child_process";
import { existsSync, mkdirSync } from "node:fs";
import { extname, join, resolve } from "node:path";
import { UsageError } from "../../../errors/index.js";
import type { Runtime } from "./runtime.js";

/** Result of running a subprocess to completion. */
export interface SpawnResult {
  readonly exitCode: number;
  readonly stdout: string;
  readonly stderr: string;
}

/** A function that runs a command to completion (injectable for tests). */
export type SpawnFn = (
  command: string,
  args: readonly string[],
  options: { readonly cwd: string; readonly env: NodeJS.ProcessEnv },
) => Promise<SpawnResult>;

/** Inputs for one synthesis run. */
export interface SynthesizeOptions {
  /** Directory containing the SDK code and stigmer.yaml. */
  readonly projectDir: string;
  /** Entry point file, relative to `projectDir`. */
  readonly entryPoint: string;
  /** Runtime inferred from the entry point extension. */
  readonly runtime: Runtime;
  /** Resolved org, injected as STIGMER_ORG_ID (empty = omit). */
  readonly orgId: string;
}

/** Injectable collaborators (defaults run real subprocesses / fs checks). */
export interface SynthesizeDeps {
  readonly spawn?: SpawnFn;
  readonly prepare?: (runtime: Runtime, projectDir: string) => void;
}

/** Outcome of synthesis: where the `.pb` files were written, plus SDK stdout. */
export interface SynthesizeResult {
  readonly outputDir: string;
  readonly stdout: string;
}

/**
 * Execute the SDK entry point and leave the synthesized `.pb` files in
 * `<projectDir>/.stigmer`. Validates inputs, runs runtime prep, spawns the
 * program with the synthesis env, and maps a non-zero exit (or spawn failure)
 * to an actionable error. The caller reads the output dir with the reader.
 */
export async function synthesize(opts: SynthesizeOptions, deps: SynthesizeDeps = {}): Promise<SynthesizeResult> {
  if (opts.projectDir === "") throw new UsageError("project directory is required");
  if (opts.entryPoint === "") throw new UsageError("entry point is required");

  const projectDir = resolve(opts.projectDir);
  if (!existsSync(projectDir)) throw new UsageError(`project directory not found: ${opts.projectDir}`);
  const entryAbs = join(projectDir, opts.entryPoint);
  if (!existsSync(entryAbs)) throw new UsageError(`entry point not found: ${opts.entryPoint}`);

  const outputDir = join(projectDir, ".stigmer");
  mkdirSync(outputDir, { recursive: true });

  (deps.prepare ?? defaultPrepare)(opts.runtime, projectDir);

  // Pass the ABSOLUTE entry path: `npx` resolves a relative entry against the
  // npm project root (its "local prefix"), not the spawn cwd, so a relative path
  // breaks for the node runtime. An absolute path is correct for every runtime;
  // cwd stays the project dir so the program's own relative imports resolve.
  const [command, ...args] = runtimeCommand(opts.runtime, entryAbs);
  const env: NodeJS.ProcessEnv = { ...process.env, STIGMER_OUT_DIR: outputDir };
  if (opts.orgId !== "") env.STIGMER_ORG_ID = opts.orgId;

  let result: SpawnResult;
  try {
    result = await (deps.spawn ?? defaultSpawn)(command, args, { cwd: projectDir, env });
  } catch (err) {
    throw new UsageError(formatExecutionError(opts.runtime, "", (err as Error).message));
  }
  if (result.exitCode !== 0) {
    throw new UsageError(formatExecutionError(opts.runtime, result.stderr, `exited with code ${result.exitCode}`));
  }

  return { outputDir, stdout: result.stdout };
}

/** Command + args to execute an entry point for the given runtime. */
export function runtimeCommand(runtime: Runtime, entryPoint: string): string[] {
  switch (runtime) {
    case "go":
      return ["go", "run", entryPoint];
    case "python":
      return ["python3", entryPoint];
    case "node": {
      const ext = extname(entryPoint).toLowerCase();
      // TypeScript entries run through tsx; plain JS runs on node directly.
      return ext === ".ts" || ext === ".tsx" || ext === ".mts" ? ["npx", "tsx", entryPoint] : ["node", entryPoint];
    }
  }
}

// Lightweight, runtime-specific readiness checks (no subprocess). Unlike Go we
// do NOT auto-run `go mod tidy`; the entry-point run surfaces missing deps and
// the user resolves them — keeping the single subprocess the synthesis itself.
function defaultPrepare(runtime: Runtime, projectDir: string): void {
  if (runtime === "node") {
    if (!existsSync(join(projectDir, "package.json"))) {
      throw new UsageError("package.json not found: run 'npm init' to initialize a Node.js project");
    }
    if (!existsSync(join(projectDir, "node_modules"))) {
      throw new UsageError("node_modules not found: run 'npm install' to install dependencies");
    }
  } else if (runtime === "go" && !existsSync(join(projectDir, "go.mod"))) {
    throw new UsageError("go.mod not found: run 'go mod init' to initialize a Go module");
  }
}

// Real subprocess: collects stdout/stderr, resolves with the exit code.
const defaultSpawn: SpawnFn = (command, args, options) =>
  new Promise<SpawnResult>((resolvePromise, rejectPromise) => {
    const child = spawn(command, [...args], { cwd: options.cwd, env: options.env });
    let stdout = "";
    let stderr = "";
    child.stdout?.on("data", (chunk) => {
      stdout += chunk.toString();
    });
    child.stderr?.on("data", (chunk) => {
      stderr += chunk.toString();
    });
    child.on("error", (err) => rejectPromise(err));
    child.on("close", (code) => resolvePromise({ exitCode: code ?? 0, stdout, stderr }));
  });

// Actionable failure message: truncated stderr + runtime-specific guidance.
// Port of Go's formatExecutionError.
function formatExecutionError(runtime: Runtime, stderr: string, fallback: string): string {
  let detail = stderr.length > 800 ? `${stderr.slice(0, 800)}\n... (truncated)` : stderr;
  if (detail === "") detail = fallback;

  const guidance =
    runtime === "go"
      ? "Check for compile errors above. Run 'go build' to see full error output."
      : runtime === "python"
        ? "If you see import errors, install your dependencies (e.g. pip install -r requirements.txt)."
        : "If you see module errors, run 'npm install' to install dependencies.";

  return `SDK synthesis failed:\n${detail}\n\n${guidance}`;
}
